plugins/pipewire/PipeWireDevice.cpp

```diff
@@ -0,0 +1,376 @@
+/*******************************************************************************
+ * Copyright 2009-2024 Jörg Müller
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	 http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+#include "PipeWireDevice.h"
+
+#include <spa/param/audio/format-utils.h>
+
+#include "Exception.h"
+#include "IReader.h"
+#include "PipeWireLibrary.h"
+
+#include "devices/DeviceManager.h"
+#include "devices/IDeviceFactory.h"
+
+AUD_NAMESPACE_BEGIN
+
+PipeWireDevice::PipeWireSynchronizer::PipeWireSynchronizer(PipeWireDevice* device) : m_device(device)
+{
+}
+
+void PipeWireDevice::PipeWireSynchronizer::update_tick_start()
+{
+	pw_time tm;
+	AUD_pw_stream_get_time_n(m_device->m_stream, &tm, sizeof(tm));
+	m_tick_start = tm.ticks;
+	m_seek_pos = m_timeline_pos;
+}
+
+void PipeWireDevice::PipeWireSynchronizer::seek(std::shared_ptr<IHandle> handle, double time)
+{
+	pw_time tm;
+	AUD_pw_stream_get_time_n(m_device->m_stream, &tm, sizeof(tm));
+	m_tick_start = tm.ticks;
+	m_seek_pos = m_timeline_pos = time;
+	{
+		std::unique_lock<std::mutex> lock(m_device->m_mixingLock);
+		m_device->m_clear_ringbuffer = true;
+	}
```

Comment:
Please remove the functionality to clear the ring buffer. The reason is that this can cause skips in sounds that are playing back but not being synchronized. This is especially bad for live sources like microphones as there would not even be any future data available for them.

Reply:
I had to clear it as it would otherwise produce sound glitches when starting playback after pausing and seeking.
(Because old audio data would be played back before the new data could be queued)

Reply:
It's causing other issues though. I had to solve the same problem for Pulseaudio, please check how it's done there. In a nutshell: you just keep playing until the ring buffer is empty and are sure all samples from it have been played and then stop.